src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,416 @@
+}
+
+std::string TDigest::internalValueFromCentroid(const Centroid& centroid) {
+  std::string value;
+  PutDouble(&value, centroid.weight);
+  return value;
+}
+
+rocksdb::Status TDigest::decodeCentroidFromKeyValue(const rocksdb::Slice& key, const rocksdb::Slice& value,
+                                                    Centroid* centroid) const {
+  InternalKey ikey(key, storage_->IsSlotIdEncoded());
+  auto subkey = ikey.GetSubKey();
+  auto type_flg = static_cast<uint8_t>(SegmentType::kGuardFlag);
+  GetFixed8(&subkey, &type_flg);
+  if (static_cast<SegmentType>(type_flg) != SegmentType::kCentroids) {
+    return rocksdb::Status::Corruption(fmt::format("corrupted tdigest centroid key type: {}", type_flg));
+  }
+  GetDouble(&subkey, &centroid->mean);
+  rocksdb::Slice value_slice = value;  // GetDouble needs a mutable pointer of slice
+  GetDouble(&value_slice, &centroid->weight);
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status TDigest::appendBuffer(engine::Context& ctx, ObserverOrUniquePtr<rocksdb::WriteBatchBase>& batch,
+                                      const std::string& ns_key, const std::vector<double>& inputs,
+                                      TDigestMetadata* metadata) {
+  // must guard by lock
+  auto buffer_key = internalBufferKey(ns_key, *metadata);
+  std::string buffer_value;
+  if (auto status = storage_->Get(ctx, ctx.GetReadOptions(), cf_handle_, buffer_key, &buffer_value);
+      !status.ok() && !status.IsNotFound()) {
+    return status;
+  }
+
+  for (auto item : inputs) {
+    PutDouble(&buffer_value, item);
+  }
+
+  if (auto status = batch->Put(cf_handle_, buffer_key, buffer_value); !status.ok()) {
+    return status;
+  }
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status TDigest::dumpCentroidsAndBuffer(engine::Context& ctx, const std::string& ns_key,
+                                                const TDigestMetadata& metadata, std::vector<Centroid>* centroids,
+                                                std::vector<double>* buffer,
+                                                ObserverOrUniquePtr<rocksdb::WriteBatchBase>* clean_after_dump_batch) {
+  if (buffer != nullptr) {
+    buffer->clear();
+    buffer->reserve(metadata.unmerged_nodes);
+    auto buffer_key = internalBufferKey(ns_key, metadata);
+    std::string buffer_value;
+    auto status = storage_->Get(ctx, ctx.GetReadOptions(), cf_handle_, buffer_key, &buffer_value);
+    if (!status.ok() && !status.IsNotFound()) {
+      return status;
+    }
+
+    if (status.ok()) {
+      rocksdb::Slice buffer_slice = buffer_value;
+      for (uint64_t i = 0; i < metadata.unmerged_nodes; ++i) {
+        double tmp_value = std::numeric_limits<double>::quiet_NaN();
+        if (!GetDouble(&buffer_slice, &tmp_value)) {
+          return rocksdb::Status::Corruption(
+              fmt::format("metadata has {} records, but get {} failed", metadata.unmerged_nodes, i));
```

Comment:
should we mention this is in buffer?

Reply:
Hi @mapleFU ,

Got it! 😊
I have moved the messages to log.

Best Regards,
Edward